File: jsr_331/FurnitureMoving.java
package org.jcp.jsr331.hakan;

/**
 *
 * Furniture Moving problem in choco v2.
 *
 * Problem from Marriott & Stuckey: 'Programming with constraints', page  112f
 *
 * Feature: testing cumulative.
 *
 * Note: JSR-331 don't support cumulative as of writing, so this model
 *       includes a decomposition of this global constraint.       
 *
 * Compare with the following models:
 * - Choco: http://www.hakank.org/choco/FurnitureMoving.java
 * - Comet: http://www.hakank.org/comet/furniture_moving.co
 * - ECLiPSE: http://www.hakank.org/eclipse/furniture_moving.ecl
 * - Gecode: http://www.hakank.org/gecode/furniture_moving.cpp
 * - Google CP Solver: http://www.hakank.org/google_or_tools/furniture_moving.py
 * - JaCoP: http://www.hakank.org/JaCoP/FurnitureMoving.java
 * - MiniZinc: http://www.hakank.org/minizinc/furniture_moving.mzn
 * - SICStus: http://www.hakank.org/sicstus/furniture_moving.pl
 * - Zinc: http://www.hakank.org/minizinc/furniture_moving.zinc
 *
 * Model by Hakan Kjellerstrand (hakank@bonetmail.com)
 * Also see http://www.hakank.org/jsr_331/
 */
import javax.constraints.*;

import java.io.*;
import java.util.*;
import java.text.*;

public class FurnitureMoving {

    Problem p = ProblemFactory.newProblem("FurnitureMoving");

    /**
     *
     * Experimental decomposition of cumulative.
     *
     * Inspired by the MiniZinc implementation:
     * http://www.g12.csse.unimelb.edu.au/wiki/doku.php?id=g12:zinc:lib:minizinc:std:cumulative.mzn&s[]=cumulative
     * The MiniZinc decomposition is discussed in the paper:
     * A. Schutt, T. Feydy, P.J. Stuckey, and M. G. Wallace.
     * 'Why cumulative decomposition is not as bad as it sounds.'
     * Download:
     * http://www.cs.mu.oz.au/%7Epjs/rcpsp/papers/cp09-cu.pdf
     * http://www.cs.mu.oz.au/%7Epjs/rcpsp/cumu_lazyfd.pdf
     *
     * Parameters:
     *   s: start_times    assumption: array of IntVar
     *   d: durations      assumption: array of int
     *   r: resources      assumption: array of int
     *   b: resource limit assumption: IntVar or int
     */
    public void my_cumulative(Var[] s, int[] d, int[] r, Var b) {
        int n = s.length;

        int times_min = 999999;
        int times_max = 0;
        for(int i = 0; i < n; i++) {
            int s_min = s[i].getMin();
            int s_max = s[i].getMax();
            
            if (s_max > times_max) {
                times_max = s_max;
            }
            if (s_min < times_min) {
                times_min = s_min;
            }
        }

        int d_max = 0;
        for(int i = 0; i < n; i++) {
            if (d[i] > d_max) {
                d_max = d[i];
            }
        }        
        times_max = times_max + d_max;
        
        for(int t = times_min; t <= times_max; t++) {
            ArrayList<Var> bb = new ArrayList<Var>();
            for(int i = 0; i < n; i++) {
                Var c1 = p.linear(s[i], "<=", t).asBool();
                Var c2 = p.linear(s[i].plus(d[i]), ">", t).asBool();
                bb.add(c1.multiply(c2).multiply(r[i]));
            }
            p.post(p.sum(bb.toArray(new Var[1])), "<=", b);
        }

        // sanity check: ensure that b < sum(r)
        int sum_r = 0;
        for(int i = 0; i < n; i++) {
            sum_r += r[i];
        }
        p.post(b, "<=", sum_r);

    }

    public static void main(String[] args) {
        FurnitureMoving p = new FurnitureMoving();
        p.define();
        p.solve();
    }
    

    // Problem definition    
    public void define() {

        Var numPersons = p.variable("numPersons", 1, 5); // will be minimized
        int maxTime    = 60;
        
        // Start times
        Var Sp = p.variable("Sp", 0, maxTime); // Piano
        Var Sc = p.variable("Sc", 0, maxTime); // Chair 
        Var Sb = p.variable("Sb", 0, maxTime); // Bed
        Var St = p.variable("St", 0, maxTime); // Table
        Var sumStartTimes = p.variable("SumStartTimes", 0, 1000);

        Var[] starts = {Sp,Sc,Sb,St};
        p.post(p.sum(starts), "=", sumStartTimes);
        
        Var[] endTimes      = new Var[4];
        int[] durations = {30,10,15,15}; // duration of task
        int[] resources = {3,1,3,2};     // resources: num persons required for each task
        for (int i = 0; i < durations.length; i++) {
            endTimes[i]  = p.variable("end-"+i, 0, maxTime);
            p.post(starts[i].plus(durations[i]), "=",  endTimes[i]);
        }
    
        my_cumulative(starts, durations, resources, numPersons);
        
    }
    
    
    public void solve() {
        //
        // search
        //
        Solver solver = p.getSolver();
        
        SearchStrategy strategy = solver.getSearchStrategy();
        
        // strategy.setVarSelectorType(VarSelectorType.INPUT_ORDER);
        // strategy.setVarSelectorType(VarSelectorType.MIN_VALUE);
        // strategy.setVarSelectorType(VarSelectorType.MAX_VALUE);
        strategy.setVarSelectorType(VarSelectorType.MIN_DOMAIN);
        // strategy.setVarSelectorType(VarSelectorType.MIN_DOMAIN_MIN_VALUE);
        // strategy.setVarSelectorType(VarSelectorType.MIN_DOMAIN_RANDOM);
        // strategy.setVarSelectorType(VarSelectorType.RANDOM);
        // strategy.setVarSelectorType(VarSelectorType.MIN_DOMAIN_MAX_DEGREE);
        // strategy.setVarSelectorType(VarSelectorType.MIN_DOMAIN_OVER_DEGREE);
        // strategy.setVarSelectorType(VarSelectorType.MIN_DOMAIN_OVER_WEIGHTED_DEGREE);
        // strategy.setVarSelectorType(VarSelectorType.MAX_WEIGHTED_DEGREE);
        // strategy.setVarSelectorType(VarSelectorType.MAX_IMPACT);
        // strategy.setVarSelectorType(VarSelectorType.MAX_DEGREE);
        // strategy.setVarSelectorType(VarSelectorType.MAX_REGRET);
        
        
        
        
        // strategy.setValueSelectorType(ValueSelectorType.IN_DOMAIN);
        strategy.setValueSelectorType(ValueSelectorType.MIN);
        // strategy.setValueSelectorType(ValueSelectorType.MAX);
        // strategy.setValueSelectorType(ValueSelectorType.MIN_MAX_ALTERNATE);
        // strategy.setValueSelectorType(ValueSelectorType.MIDDLE);
        // strategy.setValueSelectorType(ValueSelectorType.MEDIAN);
        // strategy.setValueSelectorType(ValueSelectorType.RANDOM);
        // strategy.setValueSelectorType(ValueSelectorType.MIN_IMPACT);
        // strategy.setValueSelectorType(ValueSelectorType.CUSTOM);
        
        // solver.addSearchStrategy(new StrategyLogVariables(solver)); 
        // solver.traceExecution(true);        

        //
        // solve
        //

        
        Solution s = solver.findOptimalSolution(Objective.MINIMIZE, p.getVar("numPersons"));
        if (s == null) {
            p.log("No solution");
        } else {

            System.out.println("\nNumber of persons needed: " + s.getValue("numPersons") + ":");
            System.out.println(
                               "Piano: " + s.getValue("Sp") + ".." + s.getValue("end-0") + "\n" +
                               "Chair: " + s.getValue("Sc") + ".." + s.getValue("end-1") + "\n" +
                               "Bed  : " + s.getValue("Sb") + ".." + s.getValue("end-2") + "\n" + 
                               "Table: " + s.getValue("St") + ".." + s.getValue("end-3")
                               );

            System.out.println();

            // s.log();
        }
        solver.logStats();
    }

}

File: z3/furniture_moving.py
#!/usr/bin/python -u
# -*- coding: latin-1 -*-
# 
# Moving furnitures (scheduling) problem in Z3
#
# Marriott & Stukey: 'Programming with constraints', page  112f
#
# The model use an experimental decomposition of the
# global constraint cumulative. See z3_utils_hakank.py
# for the definition.
#
# 
# This Z3 model was written by Hakan Kjellerstrand (hakank@gmail.com)
# See also my Z3 page: http://hakank.org/z3/
# 
# 
from __future__ import print_function
from z3_utils_hakank import *

#
# Decompositon of cumulative.
#
# Inspired by the MiniZinc implementation:
# http://www.g12.csse.unimelb.edu.au/wiki/doku.php?id=g12:zinc:lib:minizinc:std:cumulative.mzn&s[]=cumulative
# The MiniZinc decomposition is discussed in the paper:
# A. Schutt, T. Feydy, P.J. Stuckey, and M. G. Wallace.
# 'Why cumulative decomposition is not as bad as it sounds.'
# Download:
# http://www.cs.mu.oz.au/%7Epjs/rcpsp/papers/cp09-cu.pdf
# http://www.cs.mu.oz.au/%7Epjs/rcpsp/cumu_lazyfd.pdf
#
#
# Parameters:
#
# s: start_times    assumption: array of IntVar
# d: durations      assumption: array of int
# r: resources      assumption: array of int
# b: resource limit assumption: IntVar or int
#
# Note: since I don't know how to extract the bounds of the
#       domains, both times_min and times_max1 are required
#       which is the lower/upper limits of s (the start_times)
#
# def my_cumulative(sol, s, d, r, b,times_min,times_max1):

#   tasks = [i for i in range(len(s)) if r[i] > 0 and d[i] > 0]
  
#   # how do I get the upper/lower value of a decision variable?
#   # times_min = min([s[i].Min() for i in tasks])
#   # times_max = max([s[i].Max() + max(d) for i in tasks])
#   times_max = times_max1 + max(d)
#   for t in range(times_min, times_max + 1):
#     for i in tasks:
#       sol.add(Sum([(If(s[i] <= t,1,0) * If(t < s[i] + d[i],1,0))*r[i] for i in tasks])  <= b)

#   # Somewhat experimental:
#   # This constraint is needed to contrain the upper limit of b.
#   if not isinstance(b, int):
#     sol.add(b <= sum(r))


def main():

  # sol = Solver() # 25.140s
  sol = SolverFor("QF_FD") # 1.56s
  # sol = SolverFor("QF_LIA") # 10.67s

  #
  # data
  #
  n = 4
  duration = [30, 10, 15, 15]
  demand = [3, 1, 3, 2]
  max_end_time = 160

  #
  # declare variables
  #
  start_times = [ makeIntVar(sol,"start_times[%i]" % i, 0, max_end_time) for i in range(n)]
  end_times = [ makeIntVar(sol,"end_times[%i]" % i, 0, max_end_time) for i in range(n)]  
  end_time = makeIntVar(sol,"end_time", 0, max_end_time+max(duration))

  # number of needed resources, perhaps to be minimized
  num_resources = makeIntVar(sol, "num_resources", 0, 9)

  #
  # constraints
  #
  for i in range(n):
    sol.add(end_times[i] == start_times[i] + duration[i])

  maximum(sol, end_time, end_times)

  cumulative(sol, start_times, duration, demand, num_resources,0,max_end_time)

  #
  # Some extra constraints to play with
  #

  # all tasks must end within an hour
  sol.add(end_time <= 60)

  # All tasks should start at time 0
  # for i in range(n):
  #  sol.add(start_times[i] == 0)

  # limitation of the number of people
  # sol.add(num_resources <= 4)
  sol.add(num_resources <= 5)  

  #
  # objective
  # sol.minimize(end_time)
  # sol.minimize(num_resources)

  # solution and search
  # result
  num_solutions = 0
  while sol.check() == sat:
    num_solutions += 1
    mod = sol.model()
    print("num_resources:", mod.eval(num_resources))
    print("demand       :", demand)
    print("start_times  :", [mod.eval(start_times[i]) for i in range(n)])
    print("duration     :", [duration[i] for i in range(n)])
    print("end_times    :", [mod.eval(end_times[i]) for i in range(n)])
    print("end_time     :", mod.eval(end_time))
    print()
    # getLessSolution(sol,mod,end_time)
    getLessSolution(sol,mod,num_resources+10*end_time) # "multi-objective"
    # getLessSolution(sol,mod,Sum(start_times))
    #getDifferentSolution(sol,mod,start_times,end_times)


  print()
  print("num_solutions:", num_solutions)

if __name__ == "__main__":
  main()


